Get plot data for prepostfit experiments

lpoug opened the pull request:
- Closes #418
- Allows access to the data of PrePostFit experiments through the use of `.get_plot_data()` function

___ 

Note: First time doing a proper PR so not sure if all pre-requisites are here.

<!-- readthedocs-preview causalpy start -->
----
📚 Documentation preview 📚: https://causalpy--438.org.readthedocs.build/en/438/

<!-- readthedocs-preview causalpy end -->

lpoug commented:
> Humble apologies for taking so long to getting to this PR @lpoug. I've unfortunately not had much time to spend on CausalPy as I'd have liked, but hoping to catch up with the backlog.
> 
> There are currently a couple of issues with the remote checks. I'm hoping to get these resolved in #437, at which point I'll test this out locally and give feedback if necessary before we can merge this :)

Absolutely no problem whatsoever @drbenvincent! Let me know when the time comes, I'll be around 😄

drbenvincent commented:
Hi @lpoug. I pushed some changes, can you make sure to pull the latest version?

I'll try to review this in the next few days :)

lpoug commented:
Hey there @drbenvincent. Just to be sure, are you waiting on anything on my side?

drbenvincent reviewed (changes requested):
Sorry about the slow review on this. My bad.

Overall this looks good. I've suggested some minor changes. Other than that the main thing is to update the tests to ensure this functionality remains working into the future.

Could you add new tests to `test_integration_pymc_examples.py` and  `test_integration_skl_examples.py`. I imagine we can just test that we successfully get back a dataframe from calling `result.get_plot_data` on the experiments that you've implemented so far. You could optionally test that the contents of that dataframe is as expected, e.g. has the desired columns.

In theory an ultra pedantic person might want to test that we get an exception when calling the `get_plot_data` on experiments that.

Because this PR involves additional methods, can you run `make uml`. This should update the UML diagram that we include in `CONTRIBUTING.md`

Sorry again about the latency on this review.

lpoug commented on causalpy/experiments/base.py:
```diff
@@ -78,3 +79,26 @@ def bayesian_plot(self, *args, **kwargs):
+            raise ValueError("Unsupported model type")
+
+    @abstractmethod
+    def get_plot_data_bayesian(self, *args, **kwargs):
```
Modifications made in [6a6face](https://github.com/pymc-labs/CausalPy/pull/438/commits/6a6facecfc9a34cdd6fde89b8186ff67f111d330)

I took the liberty of renaming `plot_bayesian` and `plot_ols` to `_plot_bayesian` and `_plot_ols`, respectively. I guess this should follow the same logic as for `_get_plot_data_bayesian`? Let me know if you think differently and I'll reverse these changes.

lpoug commented on causalpy/experiments/base.py:
```diff
@@ -78,3 +79,26 @@ def bayesian_plot(self, *args, **kwargs):
+        raise NotImplementedError("get_plot_data_bayesian method not yet implemented")
+
+    @abstractmethod
+    def get_plot_data_ols(self, *args, **kwargs):
```
Modifications made in [6a6face](https://github.com/pymc-labs/CausalPy/pull/438/commits/6a6facecfc9a34cdd6fde89b8186ff67f111d330) as well

drbenvincent reviewed (changes requested):
Looks good. I think we are very nearly there :) Thanks for adding in the tests.

It could be prudent to rename the `*_hdi_lower` and `*_hdi_upper` columns to include the numerical `hdi_prob` as a percentage. For example if `hdi_prob=0.8` then the columns could be labelled as `*_hdi_lower_80` and `*_hdi_upper_80`. That way there is much less scope to make a mistake like generating 80% HDI's but forgetting that and thinking that you generated 95% HDI's for example. I think that should be pretty simple to do in `_get_plot_data_bayesian` or `_get_plot_data_ols`.

Should also be pretty simple to resolve the merge conflicts - it's just the updated uml images as far as I can see.

Could you also update the docstring of the tests? At the moment they list out what is tested, so you can just flag up that it tests the functionality of `plot_data`. I'm not sure we'll carry on doing that in the future if the number of tests gets large, but let's keep up with it for the moment.

lpoug commented on causalpy/tests/test_integration_pymc_examples.py:
```diff
@@ -378,9 +379,21 @@ def test_its():
+    )
+    expected_columns = [
+        "prediction",
+        "pred_hdi_lower_94",
```
is it ok here to have fixed names with the default value of hdi_prob (i.e. 0.94)?

drbenvincent replied:
Oh, it's actually possible to get the default value like this...

```python
import arviz as az
hdi_prob = az.rcParams["stats.ci_prob"]
```

Could be good to do that, just on the off chance that the default changes in arviz.

lpoug commented:
I made the changes for dynamic naming of `*_hdi_upper` and `*_hdi_lower` and adjusted the tests accordingly (please see my comment above regarding tests). I also updated the docstrings of the tests as requested. 
See commit [44d3870](https://github.com/pymc-labs/CausalPy/pull/438/commits/44d3870730f516e5d7ec566bb76015b0c1ff3c51)

Regarding the merge conflicts, to be honest I'm not sure what I need to do on my side? Could you enlighten me, please?

Thanks!

drbenvincent commented:
Hi @lpoug. I've sorted the conflicting files and done a few small things. I'm just noticing that at the moment the rendered docs obscure the allowable args/kwargs. So we have this:

<img width="748" alt="Screenshot 2025-04-16 at 10 08 06" src="https://github.com/user-attachments/assets/56a100ae-6fe7-47ce-92b8-74b4974cf64b" />
Which doesn't give the user much hint of what they can pass.

Same situation if we look at an actual experiment class:
<img width="751" alt="Screenshot 2025-04-16 at 10 08 51" src="https://github.com/user-attachments/assets/48478171-0bd8-467b-9751-a504abe38abc" />

I will have a quick play with this to see if we can expose the parameter information to the user in the docs. I think the easiest way is to revert a previous suggestion and make `_get_plot_data_bayesian` and `_get_plot_data_ols` public methods again. I experiment and make a commit shortly.

drbenvincent commented:
Any other final changes @lpoug, or are you happy to merge this now?

drbenvincent commented:
Success with code coverage and all tests passing. Hope you don't mind that I carried this over the finish like @lpoug. Sorry the process was a little slow, it will be faster on your next PR 😀

lpoug commented:
Amazing thank you @drbenvincent! Glad to have been a part of this 😄 Looking forward for the next, even more challenging one!
